File: core/src/main/java/no/elg/infiniteBootleg/events/EventManager.java
package no.elg.infiniteBootleg.events;

import java.lang.reflect.Method;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;

public class EventManager {

  private final Map<Class<? extends Event>, Set<EventListener>> listeners;

  public EventManager() {
    listeners = new ConcurrentHashMap<>();
  }

  public void registerListener(EventListener listener) {
    for (Method method : listener.getClass().getMethods()) {
      // TODO find all method that has a event as one and only parameter and is annotated with
      // EventHandler
    }
  }
}

File: core/src/main/kotlin/no/elg/infiniteBootleg/events/EventManager.kt
package no.elg.infiniteBootleg.events

import java.util.concurrent.ConcurrentHashMap

class EventManager {
  private val listeners: Map<Class<out Event>, Set<EventListener>> = ConcurrentHashMap()

  fun registerListener(listener: EventListener) {
    for (method in listener.javaClass.methods) {
      // TODO find all method that has a event as one and only parameter and is annotated with
      // EventHandler
    }
  }
}
